src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -0,0 +1,264 @@
+using Datadog.Trace.Logging;
+using Datadog.Trace.Util;
+using Datadog.Trace.Vendors.Serilog.Events;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Abstractions;
+using Microsoft.AspNetCore.Mvc.Controllers;
+using Microsoft.AspNetCore.Mvc.Filters;
+
+namespace Datadog.Trace.DiagnosticListeners
+{
+    /// <summary>
+    /// Instruments ASP.NET Core.
+    /// <para/>
+    /// Unfortunately, ASP.NET Core only uses one <see cref="System.Diagnostics.DiagnosticListener"/> instance
+    /// for everything so we also only create one observer to ensure best performance.
+    /// <para/>
+    /// Hosting events: https://github.com/dotnet/aspnetcore/blob/master/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs
+    /// </summary>
+    internal sealed class AspNetCoreDiagnosticObserver : DiagnosticObserver
+    {
+        public const string IntegrationName = "AspNetCore";
+
+        private const string DiagnosticListenerName = "Microsoft.AspNetCore";
+        private const string ComponentName = "aspnet_core";
+        private const string HttpRequestInOperationName = "aspnet_core.request";
+        private const string NoHostSpecified = "UNKNOWN_HOST";
+
+        private static readonly Vendors.Serilog.ILogger Log = DatadogLogging.For<AspNetCoreDiagnosticObserver>();
+
+        private static readonly PropertyFetcher HttpRequestInStartHttpContextFetcher = new PropertyFetcher("HttpContext");
+        private static readonly PropertyFetcher HttpRequestInStopHttpContextFetcher = new PropertyFetcher("HttpContext");
+        private static readonly PropertyFetcher UnhandledExceptionHttpContextFetcher = new PropertyFetcher("HttpContext");
+        private static readonly PropertyFetcher UnhandledExceptionExceptionFetcher = new PropertyFetcher("Exception");
+        private static readonly PropertyFetcher BeforeActionHttpContextFetcher = new PropertyFetcher("HttpContext");
+        private static readonly PropertyFetcher BeforeActionActionDescriptorFetcher = new PropertyFetcher("ActionDescriptor");
+        private static readonly PropertyFetcher AfterOnActionActionExecutedContextFetcher = new PropertyFetcher("ActionExecutedContext");
+
+        private readonly IDatadogTracer _tracer;
+        private readonly AspNetCoreDiagnosticOptions _options;
+
+        public AspNetCoreDiagnosticObserver(IDatadogTracer tracer, AspNetCoreDiagnosticOptions options)
+            : base(tracer)
+        {
+            _tracer = tracer ?? throw new ArgumentNullException(nameof(tracer));
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        protected override string GetListenerName() => DiagnosticListenerName;
+
+        protected override void OnNext(string eventName, object arg)
+        {
+            switch (eventName)
+            {
+                case "Microsoft.AspNetCore.Hosting.HttpRequestIn.Start":
+                    OnHostingHttpRequestInStart(arg);
```

Comment:
Is it necessary to have both the `Start` and `Stop` events received for an `HttpRequest` in order to get the correct ordering of spans?

Reply:
We need to measure the time between the `Start` and `Stop` events to determine the span's duration.

Reply:
Maybe I didn't word my question very well. What I meant to ask was whether the spans would be messed up if for some reason a `Stop` event was not received. I don't know how that could happen but maybe if an exception is thrown. In that case, would a span that had started but not stopped be considered finished at that point?

Reply:
Is Stop called after `OnHostingUnhandledException` is called?
It could be good to backlog an orphaned span monitor, basically, close a span after a certain amount of time and whatever other flags make sense.

Reply:
Yes, `Stop` is always called, even if there's an unhandled exception in user code.

Reply:
That's what I wanted to know from my first question. 👍